src/stream/src/executor/backfill/snapshot_backfill/executor.rs

```diff
@@ -515,22 +482,69 @@ async fn receive_next_barrier(
         .ok_or_else(|| anyhow!("end of barrier receiver"))?)
 }
 
-#[try_stream(ok = StreamChunk, error = StreamExecutorError)]
-async fn make_snapshot_stream<'a>(
-    row_stream: impl Stream<Item = StreamExecutorResult<OwnedRow>> + 'a,
-    mut builder: DataChunkBuilder,
-) {
-    pin_mut!(row_stream);
-    while let Some(row) = row_stream.try_next().await? {
-        if let Some(data_chunk) = builder.append_one_row(row) {
-            let ops = vec![Op::Insert; data_chunk.capacity()];
-            yield StreamChunk::from_parts(ops, data_chunk);
-        }
-    }
-    if let Some(data_chunk) = builder.consume_all() {
-        let ops = vec![Op::Insert; data_chunk.capacity()];
-        yield StreamChunk::from_parts(ops, data_chunk);
-    }
+async fn make_log_stream(
+    upstream_table: &StorageTable<impl StateStore>,
+    prev_epoch: u64,
+    start_pk: Option<OwnedRow>,
+    chunk_size: usize,
+) -> StreamExecutorResult<VnodeStream<impl super::vnode_stream::BackfillRowStream>> {
+    let data_types = upstream_table.schema().data_types();
+    let start_pk = start_pk.as_ref();
+    let vnode_streams = try_join_all(upstream_table.vnodes().iter_vnodes().map(move |vnode| {
```

Comment:
I suppose we don't iterate by vnode. Which means there can be poor locality when backfilling, since we interleave results from different streams.

Can we make a note of it here, in case we need to optimize this case down the road?

Reply:
The current implementation basically have the same behavior as the stream returned from `StorageTable::iter_with_encoded_key_range`, which merges the streams of multiple vnodes with  `streams.flatten_unordered(1024)`, and also interleave results from different streams as what we do in this PR.

In the future I think we can try limit the number of vnodes that we concurrently poll on.